ludwig/schema/config_object.py

```diff
@@ -0,0 +1,413 @@
+from ludwig.schema.utils import BaseMarshmallowConfig, convert_submodules
+from ludwig.utils.backward_compatibility import upgrade_to_latest_version
+from ludwig.utils.misc_utils import set_default_value
+
+DEFAULTS_MODULES = {NAME, COLUMN, PROC_COLUMN, TYPE, TIED, DEFAULT_VALIDATION_METRIC}
+
+
+class BaseFeatureContainer:
+    """Base Feature container for input and output features."""
+
+    def to_dict(self):
+        """Method for getting a dictionary representation of the input features.
+
+        Returns:
+            Dictionary of input features specified.
+        """
+        return convert_submodules(self.__dict__)
+
+    def to_list(self):
+        """Method for getting a list representation of the input features.
+
+        Returns:
+            List of input features specified.
+        """
+        return list(convert_submodules(self.__dict__).values())
+
+    def __repr__(self):
+        return yaml.dump(self.to_dict(), sort_keys=False)
+
+
+class InputFeaturesContainer(BaseFeatureContainer):
+    """InputFeatures is a container for all input features."""
+
+    def __init__(self):
+        pass
+
+
+class OutputFeaturesContainer(BaseFeatureContainer):
+    """OutputFeatures is a container for all output features."""
+
+    def __init__(self):
+        pass
+
+
+@dataclass(repr=False)
+class Config(BaseMarshmallowConfig):
+    """This class is the implementation of the config object that replaces the need for a config dictionary
+    throughout the project."""
+
+    def __init__(self, config_dict: dict):
+
+        upgraded_config = upgrade_to_latest_version(config_dict)
+
+        self.model_type: str = MODEL_ECD
+        self.input_features: InputFeaturesContainer = copy.deepcopy(InputFeaturesContainer())
+        self.output_features: OutputFeaturesContainer = copy.deepcopy(OutputFeaturesContainer())
+        self.combiner: BaseCombinerConfig = copy.deepcopy(ConcatCombinerConfig())
+        self.trainer: BaseTrainerConfig = copy.deepcopy(ECDTrainerConfig())
+        self.preprocessing: PreprocessingConfig = copy.deepcopy(PreprocessingConfig())
+        self.defaults: DefaultsConfig = copy.deepcopy(DefaultsConfig())
+
+        # ===== Defaults =====
+        if DEFAULTS in upgraded_config:
+            self._set_attributes(self.defaults, upgraded_config[DEFAULTS])
+
+        # ===== Features =====
+        self._set_feature_column(upgraded_config)
+        self._set_proc_column(upgraded_config)
+        self._parse_features(upgraded_config[INPUT_FEATURES], INPUT_FEATURES)
+        self._parse_features(upgraded_config[OUTPUT_FEATURES], OUTPUT_FEATURES)
+
+        # ===== Model Type =====
+        if MODEL_TYPE in upgraded_config:
+            if upgraded_config[MODEL_TYPE] == MODEL_GBM:
+                self.model_type = MODEL_GBM
+                self.trainer = GBMTrainerConfig()
+                if TYPE in upgraded_config.get(TRAINER, {}):
+                    assert upgraded_config[TRAINER][TYPE] == "lightgbm_trainer"
```

Comment:
This could also be `lightgbm_ray_trainer` depending on the backend

Reply:
Good look, will make the change.